File: 15-reverse-linked-list/solution.py
# Reverse Linked List (Iterative and Recursive)
# Time: O(n), Space: O(1) iterative / O(n) recursive

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def reverseListIterative(self, head: ListNode) -> ListNode:
        prev = None
        curr = head
        
        while curr:
            next_temp = curr.next
            curr.next = prev
            prev = curr
            curr = next_temp
        
        return prev
    
    def reverseListRecursive(self, head: ListNode) -> ListNode:
        if not head or not head.next:
            return head
        
        new_head = self.reverseListRecursive(head.next)
        head.next.next = head
        head.next = None
        
        return new_head

# Test
if __name__ == "__main__":
    # Create list: 1 -> 2 -> 3
    head = ListNode(1, ListNode(2, ListNode(3)))
    sol = Solution()
    reversed_head = sol.reverseListIterative(head)
    # Print: 3 -> 2 -> 1
    curr = reversed_head
    while curr:
        print(curr.val, end=" -> " if curr.next else "\n")
        curr = curr.next

File: 15-reverse-linked-list/solution.cs
// Reverse Linked List (Iterative and Recursive)
// Time: O(n), Space: O(1) iterative / O(n) recursive

using System;

public class ListNode {
    public int val;
    public ListNode next;
    
    public ListNode(int val = 0, ListNode next = null) {
        this.val = val;
        this.next = next;
    }
}

public class Solution {
    public ListNode ReverseListIterative(ListNode head) {
        ListNode prev = null;
        ListNode curr = head;
        
        while (curr != null) {
            ListNode nextTemp = curr.next;
            curr.next = prev;
            prev = curr;
            curr = nextTemp;
        }
        
        return prev;
    }
    
    public ListNode ReverseListRecursive(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        
        ListNode newHead = ReverseListRecursive(head.next);
        head.next.next = head;
        head.next = null;
        
        return newHead;
    }
    
    public static void Main(string[] args) {
        ListNode head = new ListNode(1, new ListNode(2, new ListNode(3)));
        Solution sol = new Solution();
        ListNode reversedHead = sol.ReverseListIterative(head);
        ListNode curr = reversedHead;
        while (curr != null) {
            Console.Write(curr.val + (curr.next != null ? " -> " : "\n"));
            curr = curr.next;
        }
    }
}
